Update Inductor windows tutorial with xpu support (#3309)
pytorch/tutorials: prototype_source/inductor_windows.rst

How to use ``torch.compile`` on Windows CPU/XPU
===============================================

**Author**: `Zhaoqiong Zheng <https://github.com/ZhaoqiongZ>`_, `Xu, Han <https://github.com/xuhancn>`_

Introduction
------------

TorchInductor is the new compiler backend that compiles the FX Graphs generated by TorchDynamo into optimized C++/Triton kernels.

> This tutorial introduces the steps for utilizing TorchInductor via ``torch.compile`` on Windows CPU/XPU.

Reviewer: ```suggestion
This tutorial introduces the steps for using TorchInductor via ``torch.compile`` on Windows CPU/XPU.
```

Software Installation
---------------------

Now, we will walk you through a step-by-step tutorial for how to use ``torch.compile`` on Windows CPU/XPU.

Install a Compiler
^^^^^^^^^^^^^^^^^^

C++ compiler is required for TorchInductor optimization, let's take Microsoft Visual C++ (MSVC) as an example.

> Download and install `MSVC <https://visualstudio.microsoft.com/downloads/>`_.

Reviewer: ```suggestion
1. Download and install `MSVC <https://visualstudio.microsoft.com/downloads/>`_.
```

> During Installation, select ``Workloads`` table then ``Desktop & Mobile`` Section, check mark on ``Desktop Development with C++`` and then install.

Reviewer: ```suggestion
1. During Installation, select **Workloads** and then **Desktop & Mobile**.
1. Select a checkmark on **Desktop Development with C++** and install.
```

Author: Hi @svekars , the suggestion here will leads to an index format issue. I fixed with #3343 .
Please help review and approve.

.. image:: ../_static/img/install_msvc.png

.. note::

    Windows CPU inductor also support C++ compiler `LLVM Compiler <https://github.com/llvm/llvm-project/releases>`_ and `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler-download.html>`_ for better performance.
    Please check `Alternative Compiler for better performance on CPU <#alternative-compiler-for-better-performance>`_.

Set Up Environment
^^^^^^^^^^^^^^^^^^

> 

Reviewer: ```suggestion
Next, let's configure our environment.

```

#. Open a command line environment via cmd.exe.
#. Activate ``MSVC`` via below command::
    
    "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/Build/vcvars64.bat"

> #. Create and activate a virtualvirtual environment::

Reviewer: should there be a command on how to activate virtual env?
```suggestion
#. Create and activate a virtual environment: ::
```

#. Install `PyTorch 2.5 <https://pytorch.org/get-started/locally/>`_ or later for CPU Usage. Install PyTorch 2.7 or later refer to `Getting Started on Intel GPU <https://pytorch.org/docs/main/notes/get_start_xpu.html>`_ for XPU usage.

> #. Use TorchInductor on Windows::

Reviewer: ```suggestion
#. Here is an example of how to use TorchInductor on Windows:
.. code-block:: python
```

Author: Hi @svekars , this update cause a format issue and I have an update fix for it with #3339 
Please help review!

    import torch
    device="cpu" # or "xpu" for XPU
    def foo(x, y):
        a = torch.sin(x)
        b = torch.cos(x)
        return a + b
    opt_foo1 = torch.compile(foo)
    print(opt_foo1(torch.randn(10, 10).to(device), torch.randn(10, 10).to(device)))

> #. Output of the above example::

Reviewer: ```suggestion
#. Here is the output of the above example::
```

            [-1.3428e+00,  1.8400e-01,  1.1756e+00, -3.0654e-01,  9.7973e-01,
            1.4019e+00,  1.1886e+00, -1.9194e-01,  1.3632e+00,  1.1811e+00],
            [-7.1615e-01,  4.6622e-01,  1.2089e+00,  9.2011e-01,  1.0659e+00,
            9.0892e-01,  1.1932e+00,  1.3888e+00,  1.3898e+00,  1.3218e+00],
            [ 1.4139e+00, -1.4000e-01,  9.1192e-01,  3.0175e-01, -9.6432e-01,
            -1.0498e+00,  1.4115e+00, -9.3212e-01, -9.0964e-01,  1.0127e+00],
            [ 5.7244e-04,  1.2799e+00,  1.3595e+00,  1.0907e+00,  3.7191e-01,
            1.4062e+00,  1.3672e+00,  6.8502e-02,  8.5216e-01,  8.6046e-01]])

Alternative Compiler for better performance on CPU
--------------------------------------------------

To enhance performance for inductor on Windows CPU, you can use the Intel Compiler or LLVM Compiler. However, they rely on the runtime libraries from Microsoft Visual C++ (MSVC). Therefore, your first step should be to install MSVC.

Intel Compiler
^^^^^^^^^^^^^^

#. Download and install `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler-download.html>`_ with Windows version.

> #. Set Windows Inductor Compiler via environment variable ``set CXX=icx-cl``

Reviewer: ```suggestion
#. Set Windows Inductor Compiler via environment variable ``set CXX=icx-cl``.
```

LLVM Compiler
^^^^^^^^^^^^^

#. Download and install `LLVM Compiler <https://github.com/llvm/llvm-project/releases>`_ and choose win64 version.

> #. Set Windows Inductor Compiler via environment variable ``set CXX=clang-cl`` 

Reviewer: ```suggestion
#. Set Windows Inductor Compiler via environment variable ``set CXX=clang-cl``.
```

Conclusion
----------

> With this tutorial, we introduce how to use Inductor on Windows CPU with PyTorch 2.5 later, and on Windows XPU with PyTorch 2.7 or later. We can also use Intel Compiler or LLVM Compiler to get better performance on CPU.

Reviewer: ```suggestion
In this tutorial, we introduce how to use Inductor on Windows CPU with PyTorch 2.5 or later, and on Windows XPU with PyTorch 2.7 or later. We can also use Intel Compiler or LLVM Compiler to get better performance on CPU.
```

How to use ``torch.compile`` on Windows CPU/XPU
===============================================

**Author**: `Zhaoqiong Zheng <https://github.com/ZhaoqiongZ>`_, `Xu, Han <https://github.com/xuhancn>`_

Introduction
------------

TorchInductor is the new compiler backend that compiles the FX Graphs generated by TorchDynamo into optimized C++/Triton kernels.

This tutorial introduces the steps for utilizing TorchInductor via ``torch.compile`` on Windows CPU/XPU.

Software Installation
---------------------

Now, we will walk you through a step-by-step tutorial for how to use ``torch.compile`` on Windows CPU/XPU.

Install a Compiler
^^^^^^^^^^^^^^^^^^

> C++ compiler is required for torchinductor optimization, let's take Microsoft Visual C++ (MSVC) as an example.

Reviewer 2: keep formatting consistent, ie, TorchInductor

Author: keep formatting for all the TorchInductor

Download and install `MSVC <https://visualstudio.microsoft.com/downloads/>`_.

> During Installation, select ``Workloads`` table then ``Desktop & Mobile`` Section, check mark on ``Desktop Development with C++`` and then install.

Reviewer 2: Might be nice to have screenshots here.

Author: I've added a screenshot here

.. note::

    Windows CPU inductor also support C++ compiler `LLVM Compiler <https://github.com/llvm/llvm-project/releases>`_ and `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler-download.html>`_ for better performance.
    Please check `Alternative Compiler for better performance on CPU <#alternative-compiler-for-better-performance>`_.

> Conda Installation

Reviewer 2: Per https://github.com/pytorch/pytorch/issues/149551, Conda is no longer being used.

Author: Remove the conda installation and let user create and activate virtual environment on their own

            [ 1.4139e+00, -1.4000e-01,  9.1192e-01,  3.0175e-01, -9.6432e-01,
            -1.0498e+00,  1.4115e+00, -9.3212e-01, -9.0964e-01,  1.0127e+00],
            [ 5.7244e-04,  1.2799e+00,  1.3595e+00,  1.0907e+00,  3.7191e-01,
            1.4062e+00,  1.3672e+00,  6.8502e-02,  8.5216e-01,  8.6046e-01]])

Alternative Compiler for better performance on CPU
--------------------------------------------------

To enhance performance for inductor on Windows CPU, you can use the Intel Compiler or LLVM Compiler. However, they rely on the runtime libraries from Microsoft Visual C++ (MSVC). Therefore, your first step should be to install MSVC.

Intel Compiler
^^^^^^^^^^^^^^

#. Download and install `Intel Compiler <https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler-download.html>`_ with Windows version.
#. Set Windows Inductor Compiler via environment variable ``set CXX=icx-cl``

LLVM Compiler
^^^^^^^^^^^^^

#. Download and install `LLVM Compiler <https://github.com/llvm/llvm-project/releases>`_ and choose win64 version.
#. Set Windows Inductor Compiler via environment variable ``set CXX=clang-cl`` 

Conclusion
----------

> With this tutorial, we introduce how to use Inductor on Windows CPU with PyTorch 2.5 or later. We can use Intel Compiler or LLVM Compiler to get better performance.

Reviewer 3: Can we highlight Windows XPU with PyTorch 2.7 in the conclusion as well?

Author: sure

Reviewer 3: Thanks for your update.
